---
title: "Gottesman Bergel RNA-seq Data"
output: rmdformats::readthedown
author: Christina Fitzsimmons
date: 'Last updated 2021-06'
---

# 1. Introduction
This notebook is an analysis of the data from the M. Gottesman Lab from the Bergel data set. These data may warrant repeat analysis from the raw sequencing files. These data were obtained from Rob Robey in the M. Gottesman Lab. In the below analysis, data was analyzed using the gene counts obtained from the CCR genomics core.

```{r}
# Import the libraries we will need for this analysis
library (rmdformats) #library for nice html output of notebooks
library(DESeq2) # library with tools / objects for differential expression analysis
library(tidyverse) # library for data science operations
library(pheatmap) # pheatmap library is a set of tools that allow us to visualize the data
library(RColorBrewer) # RColorBrewer gives us colors with which to plot things
library(gplots) # library for enhanced versions of standard plots
library(genefilter) # library to filter genes from high-throughput experiments
library(biomaRt) # library to query the ensembl database
library (gprofiler2) # library for functional enrichment analysis
library (ggpubr) # library for making publication quality plots
library(ComplexHeatmap) # library for making genomic heatmaps
library(circlize) # library for ciruclar layout generation

theme_set(theme_bw())
setwd("/Users/fitzsimmonscm/Documents/Projects_and_Data/Collaborations/Gottesman/RNASeq_Bergel/")

```
# 2. Importing the data for analysis
These data were analyzed using the gene counts obtained from the CCR genomics core. Please contact the genomics core to learn about library processing step upstream of this current notebook, or to obtain raw sequencing files for your own re-analysis using different parameters. 

## 2.1 Importing the metadata and raw counts
In this section of the notebook, we will import the raw count data, the metadata (information about cell type, drug resistance, etc). There are several steps involved in this, and the comments annotate the purpose of each step in the below code: 
```{r}

# Importing the metadata for our analysis
# The metadata contains information about the samples (cell type, drug resistance, etc) that we will use in our analyses
coldata <- read.csv("~/Desktop/NCI:NIH/METTL7 Summer 2021/Tutorials/RNAseq_analysis_bergel/Gottesman_Rhomedepsin.csv",
                    row.names=1)
print(coldata)
nrow(coldata)
countdata1 <- read.delim(
  "~/Desktop/NCI:NIH/METTL7 Summer 2021/Tutorials/RNAseq_analysis_bergel/RawCountFile_rsemgenes.txt", row.names = 1)

# This converts the raw data into a matrix object and then performs a cast to integer. 
# This would be better fixed at the gene counting level in RNAseq, using a counting program that does not report fractional values
# Unfortunately, I don't have access to the original data, just the count files. 
countdata1 <- as.matrix(countdata1)
storage.mode(countdata1) <- "integer"
  
# Checking that the row names in the metadata file are the same as in the columns in the data frame
all(rownames(coldata) %in% colnames(countdata1))
all(rownames(coldata) == colnames(countdata1))

# Making sure the columns and row names are in the same order, or DESeq2 will get confused
countdata1 <- countdata1[, rownames(coldata)]
all(rownames(coldata) == colnames(countdata1))

```
## 2.2 Creating the DEseq2 dataset
This next section combines the information from our matrix and our metadata to create the DEseq2 data object. Above, we have confirmed that the metadata information and the raw data contain the same number of columns and rows, and that they are in the same order. Without the step on line 53, DEseq2 will be very confused and raise and error. 
```{r}
# Create the DESeq object, combining information from the raw counts and metadata. Here, we are specifying a group we are interested in comparing. In this case, the different drug treatments
dds <- DESeqDataSetFromMatrix(countData = countdata1,
                              colData = coldata,
                              design = ~ Treatment)
dds

```
##2.3 Cleanup of the DEseq2 dataset
In this section, we remove rows where the sum of the row is > 10 counts. This helps us to avoid extrapolating too much with low-frequency genes. We do NOT normalize in this notebook, although that is an option for future analyses. 
```{r}
# This will keep all of the rows where the count is greater than 10. This removes any zeros, which can make analysis tricky. 
nrow(dds)
dds <- dds[ rowSums(counts(dds)) >10, ]
nrow(dds)

# the below command give you the raw counts for the table you created above
counts(dds)
# this command give you the row totals for each gene
rowSums(counts(dds))
```
## 2.4 Generating some exploratory plots to visualize data prior to analysis

### 2.4.1 Generating the Variance Stabilizing, Rlog, and PCA plots
The below code will create 2 plots: a variance stabilizing plot and a rlog plot. In both plots, we are visualizing the variability of the samples. the goal of a variance stabilizing transformation is to create a plot y = f(x) where the variability of the y values is not related to their mean value. The rlog is a similar plot for small (n < 30) samples
```{r}

rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)

vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)

# This data can now be plotted into 2 scatter plots
df <- bind_rows(
  
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))
  
colnames(df)[1:2] <- c("x", "y")  

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)
vsd <- vst(dds)

```
# PCA plot section
The `plotPCA` is built into the DESeq2 data set. To call it, we must first create the variance stabilizing tranformation and an rplot (above)
```{r}
plotPCA(vsd, "Treatment")
```


### 2.4.2 Creating the heat map to visualize similarity between the samples
The heat map is similar to the PCA plot and allows us to visualize how similar or different the samples are within replicates and also to each other. In the heatmap below, you can see that 
```{r}
# In this next section of code, we create a heatmap to analyze the overall similarity between the samples
sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Reds")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors,
         show_rownames=TRUE)
```
# 3. Differential Expression Analysis with DEseq2
We can ask DESeq2 to contrast any two Treatment conditions. In the example below, we are comparing D9 vehicle to the control, although in the data analysis folder that contains this notebook, other comparisons are present. 
```{r}

# Finally, we are ready to create the DESeq2 data object and perform our analysis
dds <-DESeq(dds)

# Here, we are asking to compare 2 treatments, D9_vehicle and control
res1 <- results( dds, contrast = c("Treatment", "D9_vehicle", "control") ) 
res1

# The below command gives us information about the results data frame we just created in the above window 
mcols(res1, use.names = TRUE)
```
## 3.1 Generating the MA plot
The plot visualizes the differences between measurements taken in two samples, by transforming the data onto M (log ratio) and A (mean average) scales, then plotting these values. It is one method of visualizing genomic data. 
```{r}
# This is the MA plot
plotMA( res1, ylim = c(-12, 12) )
```
## 3.2 Heatmap of Top Genes
This heatmap shows the top 100 genes with the highest variance across the samples, plotted with dendragram clustering. This gives us information about which blocks of genes co-vary across treatments and samples. 

```{r}
# This is to plot a heat map with gene clustering
# plot is centered and scaled across all samples

topVarGenes <- head( order( rowVars( assay(rld) ), decreasing=TRUE ), 100 )

heatmap.2(assay(rld)[ topVarGenes, ], 
          scale="row",
          trace="none",
          dendrogram="column",
          labRow = FALSE,
          main="Heatmap of Highest Gene Variance",
          cexCol = 0.2 + 1/log10(50),
          col = "bluered")

```


```{r}
# This section will allow us to sort and export the data as a table
resOrdered <- res1[order(res1$padj),]
resOrderedDF <- as.data.frame(resOrdered)


# Now we're going to use the gprofiler library to assign gene names to the data. Right now, all we have is the big long ENSG ID number
resOrderedDF_v2 <- resOrderedDF %>%
  rownames_to_column('ENSG_ID')

# Get the names from a gProfiler query
resOrdered_name <- gconvert(resOrderedDF_v2$ENSG_ID, organism = "hsapiens", target = "HGNC", filter_na = F) %>%
  dplyr::select(input, name) %>%
  dplyr::rename(ENSG_ID = input, gene_name = name) # change input to ENSG_ID

# Join names dataframe to resOrdered df by the common column of ENSG_ID
res_export <- inner_join(resOrdered_name, resOrderedDF_v2, by = "ENSG_ID")

# Export the dataframe to a csv file for downstream analysis
write.csv(res_export, file="/Users/fitzsimmonscm/Documents/Projects_and_Data/Collaborations/Gottesman/RNASeq_Bergel/2019.03.16_Rhomedepsin_RNASeq_D9res_D9treatvscontrol.csv")
```

```{r}
ggplot(data= resOrderedDF_v2, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(aes(color=padj < 0.05)) + ggtitle("D9 Vehicle vs Control") +xlab("log2 Fold Change") + ylab("-log10 padj")
```
```{r}
# Finally, we are ready to create the DESeq2 data object and perform our analysis
ddsD9res <-DESeq(dds)

# Here, we are asking to compare 2 treatments, D9_vehicle and control
resD9res <- results( dds, contrast = c("Treatment", "D9res_D9treat", "control") ) 
resD9res

# The below command gives us information about the results data frame we just created in the above window 
mcols(resD9res, use.names = TRUE)

# This section will allow us to sort and export the data as a table
resOrderedD9res <- res1[order(res1$padj),]
resOrderedDFD9res <- as.data.frame(resOrderedD9res)


# Now we're going to use the gprofiler library to assign gene names to the data. Right now, all we have is the big long ENSG ID number
resOrderedDF_v2_D9res <- resOrderedDFD9res %>%
  rownames_to_column('ENSG_ID')
```


```{r}
ggplot(data= resOrderedDF_v2_D9res, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(aes(color=padj < 0.05)) + ggtitle("D9res_D9treat vs Control") +xlab("log2 Fold Change") + ylab("-log10 padj")
```

