---
title: "Ryan MCF7 data Upload"
output: html_document
author: Ryan Gensler
---

```{r}

# Import the libraries we will need for this analysis
library (rmdformats) #library for nice html output of notebooks
library(DESeq2) # library with tools / objects for differential expression analysis
library(tidyverse) # library for data science operations
library(pheatmap) # pheatmap library is a set of tools that allow us to visualize the data
library(RColorBrewer) # RColorBrewer gives us colors with which to plot things
library(gplots) # library for enhanced versions of standard plots
library(genefilter) # library to filter genes from high-throughput experiments
library(biomaRt) # library to query the ensembl database
library (gprofiler2) # library for functional enrichment analysis
library (ggpubr) # library for making publication quality plots
library(pasilla)
#library(ComplexHeatmap) # library for making genomic heatmaps
library(circlize) # library for ciruclar layout generation

#set working directory
theme_set(theme_bw())
setwd("/Users/ryangensler/Documents/GitHub/METTL7/Summer_2021/Ryan/MCF7")

# make directory an object
directory <- "/Users/ryangensler/Documents/GitHub/METTL7/Summer_2021/Ryan/MCF7"
  
# 2. Pointing to HTSeq data
sampleFiles <- grep("sorted", list.files(directory),value=TRUE)

#3. Setting the conditions
condition <- c('MCF7_C-1.sorted.rmdup.count', 'MCF7_C-2.sorted.rmdup.count', 'MCF7_C-3.sorted.rmdup.count','MCF7_M300_1.sorted.rmdup.count', 'MCF7_M300_2.sorted.rmdup.count', 'MCF7_M300_3.sorted.rmdup.count')

#4 Creating the sample table 
sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = condition)
#unsure why the tilda needs to be in front of 'condition'
ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = directory,
                                       design = ~ condition)


```

